cupyx/scipy/sparse/index.py

```diff
@@ -0,0 +1,865 @@
+"""Indexing mixin for sparse matrix classes.
+"""
+import math
+from .sputils import isintlike
+
+from numpy import integer
+from numpy import dtype
+
+import cupy
+from cupy import core
+
+
+INT_TYPES = (int, integer)
+
+_int32_dtype = dtype('int32')
+_int64_dtype = dtype('int64')
+_float32_dtype = dtype('float32')
+_float64_dtype = dtype('float64')
+_complex64_dtype = dtype('complex64')
+_complex128_dtype = dtype('complex128')
+
+_module_options = ('--std=c++11',)
+
+
+def _broadcast_arrays(a, b):
+    """
+    Same as cupy.broadcast_arrays(a, b) but old writeability rules.
+    NumPy >= 1.17.0 transitions broadcast_arrays to return
+    read-only arrays. Set writeability explicitly to avoid warnings.
+    Retain the old writeability rules, as our Cython code assumes
+    the old behavior.
+    """
+    x, y = cupy.broadcast_arrays(a, b)
+
+    # Writeable doesn't seem to exist on flags in cupy
+    # x.flags.writeable = a.flags.writeable
+    # y.flags.writeable = b.flags.writeable
+    return x, y
+
+
+_bin_col_offsets_ker_types = {dtype('int32'): 'bin_col_offsets_ker_str<int>'}
+_bin_col_offsets_ker = core.RawModule(code="""
+    template<typename I>
+    __global__ void bin_col_offsets_ker_str(I n_idx,
+                                            I *col_idxs,
+                                            I *col_offsets) {
+
+        // Get the index of the thread
+        I jj = blockIdx.x * blockDim.x + threadIdx.x;
+
+        if(jj < n_idx) {
+            const I j = col_idxs[jj];
+            atomicAdd(col_offsets+j, 1);
+        }
+    }
+""", options=_module_options, name_expressions=tuple(
+    _bin_col_offsets_ker_types.values()))
+
+
+def _bin_col_offsets(n_idx, col_ids, col_offsets, tpb=32):
+    grid = math.ceil(n_idx / tpb)
+    kernel = _bin_col_offsets_ker.get_function(
+        _bin_col_offsets_ker_types[_int32_dtype])
+    kernel((grid,), (tpb,), (n_idx, col_ids, col_offsets))
+
+
+def _csr_column_index1_indptr(col_offsets, Ap, Aj):
```

Comment:
Perhaps this function can be replaced with `cupy.searchsorted`. How do you think?

Reply:
I'm happy to give it a try, though I'm having trouble seeing it would be used to directly compute the output indptr. Can you share more detail on how this would look?

Reply:
I don't know the alternative implementation now, but anyway we should find a better algorithm to avoid `n_col` size array creation in L60 at least. `n_col` value tends to be huge.
https://github.com/cupy/cupy/pull/3486/files#diff-f723ef8036e0c67c8733d4eb7f7b6d87R60

Reply:
I think it's also OK to compute the result with the dense format and convert it to the sparse format. The number of nonzero values in the result cannot be found without host-device synchronization, and all values may nonzero in worst case.

Reply:
In that worst case, a sparse format is going to add an additional array of size `n_cols` if we convert to a sparse format. I'm wondering if we could use the sparsity (number of unique col idxs vs n_col) to determine whether it's worthwhile to convert to sparse.

For example, I don't believe anything is lost with the current algorithm if the number of unique col_idxs is very close to n_cols. However, I agree that having it in dense form sounds l like a total waste if `col_offsets` is extremely sparse. Maybe we could choose a threshold and use the current code path if the ratio of unique col_idxs to n_col is something like 90% or more?

Reply:
I mean that we cannot create `new_data` smaller than the size of indices (for example, in sliceXarray indexing case, it is the product of the size of the slice and the size of array) without looking the actual data in the input arrays. Of course, we can create `new_data` smaller than `n_col` size if the size of indices is smaller than `n_col`.

Reply:
Spent some time on this one today to see where I could get. I was able to build the `column_offsets` such that it only contains the counts for the unique (and sorted) elements in `col_idxs`. One option for populating the output indices and data arrays could be to filter the input indices up front for only the matching columns, change them to their indices in the `col_idxs`, and use the smaller `col_offsets` array to determine the repeating values. This is proving to be pretty challenging to do with only a couple operations though and would require at least 1 additional copy of the input indices array.

I'm also thinking we should consider parallelizing the updates over the `col_idx` array in when it is much larger than the number of rows and number of columns.

Reply:
@asi1024 I tried several different ideas for shrinking the `col_offsets` array but they all seemed to result in having several copies of the column indices and/or the index array. The big challenge there is the need to support duplicate columns in the index. I'm at a loss. 

I've been trying to keep the performance to match the constant-time lookup that we get by keeping `col_offsets` dense. Alternatively, I suppose in the case where N is very very large, we could resort to a binary search in the `csr_column_index2` kernel. I'd also be happy to try any ideas you might have.

Reply:
I managed to find a solution that at least passes all the tests. It still needs profiling and further optimization but thus far it requires the following additional memory:
1. A mask of the input indices array (size input nnz).
2. An array of size n_indices containing the occurrence frequencies for each column in the index
3. An array that maps the columns in the mask to the array of frequencies. 

The array in 3 is necessary because it's possible the index returned from `cupy.unique` is not monotonically increasing order. Next week I'll see if I can find a way to remove this.